python/cudf_polars/cudf_polars/experimental/fusion.py

```diff
@@ -0,0 +1,187 @@
+# SPDX-License-Identifier: Apache-2.0
+"""IR node fusion."""
+
+from __future__ import annotations
+
+import operator
+from collections import defaultdict
+from functools import reduce
+from typing import TYPE_CHECKING, Any
+
+from cudf_polars.dsl.ir import (
+    IR,
+    Cache,
+    Filter,
+    GroupBy,
+    HStack,
+    MapFunction,
+    Projection,
+    Select,
+    Union,
+)
+from cudf_polars.dsl.traversal import CachingVisitor, traversal
+from cudf_polars.experimental.dispatch import generate_ir_tasks
+from cudf_polars.experimental.io import Scan, SplitScan
+
+if TYPE_CHECKING:
+    from collections.abc import Callable, MutableMapping, Sequence
+    from typing import Any
+
+    from cudf_polars.containers import DataFrame
+    from cudf_polars.experimental.base import PartitionInfo
+    from cudf_polars.experimental.dispatch import LowerIRTransformer
+    from cudf_polars.typing import Schema
+
+
+class Fused(IR):
+    """Fused IR node."""
+
+    __slots__ = ("fused_io", "subnodes")
+    _non_child = ("schema", "subnodes", "fused_io")
+    subnodes: tuple[IR, ...]
+    """Fused sub-nodes."""
+    fused_io: Union | None
+    """Fused IO node."""
+
+    def __init__(
+        self,
+        schema: Schema,
+        subnodes: tuple[IR, ...],
+        fused_io: Union | None,
+        *children: IR,
+    ):
+        self.schema = schema
+        self.subnodes = subnodes
+        self.fused_io = fused_io
+        self._non_child_args = (
+            [node.do_evaluate for node in subnodes],
+            [tuple(node._non_child_args) for node in self.subnodes],
+            [],
+            [],
+        )
+        self.children = children
+
+    @classmethod
+    def do_evaluate(
+        cls,
+        funcs: Sequence[Callable],
+        subargs: Sequence[Sequence[Any]],
+        io_funcs: Sequence[Callable],
+        io_args: Sequence[Any],
+        *children: DataFrame,
+    ) -> DataFrame:
+        """Evaluate and return a dataframe."""
+        if io_funcs:
+            children = (io_funcs[0](*io_args),)
+        for func, args in zip(funcs, subargs, strict=True):
+            children = (func(*args, *children),)
+        return children[0]
+
+
+def _fuse_ir_node(
+    ir: IR, rec: LowerIRTransformer
+) -> tuple[IR, MutableMapping[IR, PartitionInfo]]:
+    partition_info: MutableMapping[IR, PartitionInfo] = {}
+    children = ()
+    if ir.children:
+        children, _partition_info = zip(*(rec(c) for c in ir.children), strict=True)
+        partition_info = reduce(operator.or_, _partition_info)
+
+    new_node: IR
+    if ir in rec.state["fusible"]:
+        if isinstance(ir, Union):
+            new_node = Fused(ir.schema, (), ir)
+        elif len(children) == 1 and isinstance(children[0], Fused):
+            (child,) = children
+            grandchildren = child.children
+            new_node = Fused(
+                ir.schema, (*child.subnodes, ir), child.fused_io, *grandchildren
+            )
```

Comment:
So this is kind of "tricking" the task generation routine into doing partition-wise operations without generating new tasks/intermediates.

Reply:
Trying to understand how this works. I think this is trying to fuse things like:
```
HStack(
   Select(A, [1, 2]),
   Select(A, [3, 4]),
   Select(A, [5, 6]),
)
```

So we only generate a "task" for evaluating `A` and then for a "fused" expression.

But I think this rewrites to:

```
Fused(
  (Select12, Select34, Select56, HStack),
  A
)
```
But it seems like the evaluation rule would produce `HStack(Select56(Select34(Select12(A))))`. What am I missing?

Reply:
We are actually targeting simple sequential IR-node strings. `HStack` is easy to fuse, because it **can't** be used in the way you describe here (maybe you're thinking of `HConcat`?).